### ISSUE TITLE
GatsbyImage not working with JavaScript disabled via CSP (uBlock Origin, Noscript addons)

### ISSUE THREAD

**Issue by yaroslav** [gatsbyjs/gatsby#38065]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

It seems that while GatsbyImage is designed to have the `<noscript>` tag support and work with both JavaScript enabled or disabled, this is not always the case.

Many prosumers seem to use uBlock Origin (https://github.com/gorhill/uBlock/issues/308) or NoScript for blocking JavaScript for a single specific website. They seem to work by setting policies for a specific page that completely block JavaScript. However, with that blocking enabled, GatsbyImage images stop showing.

### Reproduction Link

gatsbyjs.com

### Steps to Reproduce

It works:

1. Use Firefox Stable or Developer Edition
2. Go to `about:config` and set `javascript.enabled` to `false`.
3. Refresh the build and see that images prepared via GatsbyImage are WORKING (positive).

It does not, however:

1. Get Firefox (Chrome works too).
2. Go to gatsbyjs.com.
3. Notice that the Gatsby + Netlify image banner is there and looking good.
4. Install the NoScript extension, or
5. Install uBlock Origin, and disable JavaScript for that website (click extension icon, click the `</>` looking icon, reload the page).
6. After refreshing the page, see that the banner completely disappears.

The issue is reproduceable for me on both production Gatsby websites (Cloud and not-cloud) as well as on local `gatsby serve` builds.

### Expected Result

Images still show up using what is in `<noscript>` tag.

### Actual Result

Despite `<noscript>` getting used by the browser, the image in that tag seems to have the `opacity` or `0`, so all I get is an empty placeholder.

### Environment

```shell
Latest Firefox (Stable or Developer Edition)
Latest Google Chrome stable
with NoScript (https://noscript.net/) or uBlock Origin (https://ublockorigin.com/)
```

**Comment by simevidas** [comment]
I use the uBlock Origin extension to disable JavaScript for specific websites in my desktop browser. I can reproduce the issue on https://gatsbyjs.com.

<img width="1278" alt="Screenshot 2023-05-07 at 00 01 22" src="https://user-images.githubusercontent.com/716405/236648221-f5fcab79-2858-4fc6-9606-a859ae809253.png">

**Comment by LekoArts** [comment]
Hi, thanks for the issue!

Unfortunately that is out of our control and I recommend opening an issue on the uBlock repository.
When JS is disabled, we apply this CSS in the `<head>`:

```html
<noscript>
    <style>
        .gatsby-image-wrapper noscript [data-main-image] {
            opacity: 1!important
        }

        .gatsby-image-wrapper [data-placeholder-image] {
            opacity: 0!important
        }
    </style>
</noscript>
```

This way the image is shown inside the other `<noscript>` tag.

For some reason uBlock wraps this with a `<span>` if you apply this "JS disabled" option:

```html
<span style="display: inline !important;"><style>.gatsby-image-wrapper noscript [data-main-image]{opacity:1!important}.gatsby-image-wrapper [data-placeholder-image]{opacity:0!important}</style></span>
```

That's not valid HTML and thus not executed. Nothing we can do on our side.

**Comment by yaroslav** [comment]
Thanks for checking it out, @LekoArts!

**Comment by simevidas** [comment]
Why do you have `noscript` in the CSS selector inside the `<noscript>` element? When you place `<style>` inside `<noscript>`, that style is only applied to the page when JavaScript is disabled. You don’t need to further qualify the selector with `noscript`.

uBlock Origin replaces `<noscript>` with a `<span>` in order to make the no-JS content active when the user disables JavaScript via uBlock Origin. It’s a workaround because there is no standard way to activate `<noscript>` from a browser extension.

Please try removing the unnecessary `noscript` from the selector, and check if that fixes this issue.

**Comment by simevidas** [comment]
@yaroslav uBO replaces `<noscript>` with `<span>`. This is not done erroneously, but on purpose. uBO does this because there is no standard way to activate `<noscript>` content from a browser extension.

As a result, when a user disables JS for a website via uBO, there will be no `<noscript>` elements on the page. 

In order to be compatible with such users, a website cannot rely on the existence of `<noscript>` elements on the page.

**Comment by yaroslav** [comment]
> there is no standard way

The standard way to handle clients with no JavaScript available or disable exists and is very thoroughly documented.

https://html.spec.whatwg.org/multipage/scripting.html#the-noscript-element

Are you suggesting to replace an actual standard with the support for a very specific browser add-on where the JavaScript blocking DOM modifications are not even documented?

**Comment by simevidas** [comment]
I’m suggesting to remove the unnecessary `noscript` from the selector, and check if that fixes this issue.

**Comment by LekoArts** [comment]
The `noscript` tag is there for a reason. With JS enabled we change the opacity like so:

https://github.com/gatsbyjs/gatsby/blob/22394b9a3365c8852a764b1759a96741848ccde8/packages/gatsby-plugin-image/src/gatsby-ssr.tsx#L103-L108

If we'd change the the selector, all images would be visible directly, without a blur-up effect. Thus the `noscript` tag is necessary.

**Comment by krystian3w** [comment]
test:

```css
.gatsby-image-wrapper :is(noscript, span[style="display: inline !important;"]) [data-main-image]{opacity:1!important}
.gatsby-image-wrapper [data-placeholder-image]{opacity:0!important}
```

of course this cuts off support of old browsers: https://caniuse.com/css-matches-pseudo The worst is using `:matches` for some 3 Opera releases.

Rather, the project is not so crazy that it would add to the html code by itself `style="display: inline !important;"`.

The idea of NoScript detection I don't have - as they do not use the identical method as uBo.

**Comment by simevidas** [comment]
@LekoArts If we remove `noscript` from the selector, we get this:

```html
<noscript>
    <style>
        .gatsby-image-wrapper [data-main-image] {
            opacity: 1!important
        }

        .gatsby-image-wrapper [data-placeholder-image] {
            opacity: 0!important
        }
    </style>
</noscript>
```

There are two scenarios:

1. If JavaScript is enabled, the browser will ignore the `<noscript>` element, so the contained styles will be ignored too. Therefore, `opacity: 1` will *not* apply to the image, so the image will *not* be rendered immediately. Then, the JavaScript code will animate the image into view.

2. If JavaScript is disabled, the browser (or browser extension) will activate the `<noscript>` element, and the contained styles will apply. Therefore, `opacity: 1` will apply to the image, so the image will be rendered immediately. This is what we want. Render the image immediately if JavaScript is disabled.

To sum up, `noscript` in the selector is not necessary because the entire style is ignored when JavaScript is disabled. 

Please try removing `noscript` and check if it fixes the issue with uBlock Origin.
